crates/worker/src/partition_processor_manager/snapshot_task.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::path::PathBuf;
+use std::time::SystemTime;
+
+use anyhow::Context;
+use tokio::sync::{oneshot, watch};
+use tracing::{debug, warn};
+
+use restate_partition_store::snapshots::{
+    LocalPartitionSnapshot, PartitionSnapshotMetadata, SnapshotFormatVersion,
+};
+use restate_partition_store::PartitionStoreManager;
+use restate_types::identifiers::{PartitionId, SnapshotId};
+use restate_types::logs::Lsn;
+
+/// Creates a partition store snapshot along with Restate snapshot metadata.
+pub struct SnapshotPartitionTask {
+    pub cluster_name: String,
+    pub node_name: String,
+    pub partition_id: PartitionId,
+    pub snapshot_base_path: PathBuf,
+    pub partition_store_manager: PartitionStoreManager,
+    pub archived_lsn_sender: watch::Sender<Option<Lsn>>,
+    pub result_sender: oneshot::Sender<anyhow::Result<SnapshotId>>,
+}
+
+impl SnapshotPartitionTask {
+    pub async fn create_snapshot(self) -> anyhow::Result<SnapshotId> {
+        debug!(
+            partition_id = %self.partition_id,
+            "Creating partition snapshot"
+        );
+
+        let snapshot_id = SnapshotId::new();
+        let snapshot = self
+            .partition_store_manager
+            .export_partition_snapshot(self.partition_id, snapshot_id, self.snapshot_base_path)
+            .await?;
+
+        let metadata = write_snapshot_metadata_header(
+            snapshot_id,
+            self.cluster_name,
+            self.node_name,
+            self.partition_id,
+            snapshot,
+        )
+        .await?;
+
+        // todo(pavel): SnapshotRepository integration will go in here in a future PR
+
+        self.partition_store_manager
+            .update_archived_lsn(self.partition_id, metadata.min_applied_lsn)
+            .await
+            .context("Failed updating the archived LSN attribute")?;
+
+        self.archived_lsn_sender
+            .send(Some(metadata.min_applied_lsn))?;
+
+        let _ = self
+            .result_sender
+            .send(Ok(metadata.snapshot_id))
+            .inspect_err(|err| {
+                warn!(
+                    "Failed to send snapshot acknowledgement after snapshot {} was successfully created: {:?}",
+                    metadata.snapshot_id,
+                    err
+                )
+            });
```

Comment:
What happens if an error occurs before this point? We would probably drop the sender and the receiver would see a closed error w/o much more information what happened.

Reply:
Ah! I should have seen this coming, since it bit me earlier. This is now much more robustly handled, I think!